### ISSUE TITLE
Add `new control` command for creating control repository

### ISSUE THREAD

**Issue by scotje** [puppetlabs/pdk-planning#10]
_From @bittner on October 24, 2017 14:16_

PDK should support creating and extending a [control-repository](https://puppet.com/docs/pe/latest/code_management/control_repo.html), which is enabled for validation and testing using PDK (just like the modules created by PDK).

Obsoletes #289 and #332 when implemented.
See also: https://github.com/puppetlabs/control-repo/issues/52

Proposed Solution
-------------------

Ideally,

- the validation and test setup currently present in generated modules should be taken over to the proposed [Puppet Labs control-repo](https://github.com/puppetlabs/control-repo), and
- the `pdk new control` command should generate a control-repo setup that is equivalent to the proposed Puppet Labs control-repo on GitHub.

This way, whether you use PDK or just the GitHub repo as an example, you can trust to get a usable best-practice setup.

_Copied from original issue: puppetlabs/pdk#333_

**Comment by scotje** [comment]
_From @cdenneen on October 25, 2017 16:28_

@bittner except for the metadata.json check the validation will work on existing control repo so far that I can tell.
The ability to generate a control repo would definitely need to be added and skip the metadata.json check in those cases.

**Comment by scotje** [comment]
_From @bittner on October 25, 2017 22:36_

The `Gemfile` is needed to make `pdk validate` work on a control repo, see https://github.com/puppetlabs/pdk/issues/289#issuecomment-339000049.

The `Rakefile`, `.fixtures.yml` (with `links:` to the `roles` and `profiles` folders, and more dependencies specified in the `forge_modules:` section), `spec/spec_helper.rb` and specs in `spec/classes/` is needed for `pdk test unit`.

Not sure how the `examples/` would come into play. Then there are also acceptance tests (as seen in the [PSICK repo](https://github.com/example42/psick/tree/production/skeleton/spec)). If that all makes sense it must be generated, and work out of the box. :+1:

**Comment by scotje** [comment]
_From @ardrigh on May 3, 2018 4:31_

I was talking with my team about adopting PDK, and the files PDK adds when it converts modules, so that it can run the tests and validation.

Currently our control repo has 40+ internal Puppet modules, and converting them all with PDK would potentially add a huge number of duplicate files (Gemfile etc).

It would be good to know if support for testing across control repos is still on the development plan. 

I am hopeful we will eventually be able migrate our TravisCI setup to test our control repo solely using simple PDK commands

**Comment by scotje** [comment]
_From @ardrigh on May 3, 2018 5:58_

Additional question, has any thought been given to integrating the [onceover](https://github.com/dylanratcliffe/onceover) gem into PDK for control repo testing?

**Comment by scotje** [comment]
_From @Nekototori on May 8, 2018 18:2_

I'd second onceover.  PDK is to modules as onceover is to control repos (as it stands today).

**Comment by ardrigh** [comment]
Just as a comment on development of an RFC on this type of feature...

It would be good if a `pdk new role` class integrated with the `onceover` testing to create a test for the role, in the manner that classes and defines do.

This requires work on support for testing of control repos, so it's a complicated request. 

But looking it from a greenfields user perspective, the starting point for a new site should hopefully be defining a role and some profiles, and have instant feedback that something is working because a test passed successfully.

**Comment by reidmv** [comment]
Language comment: the term "control" by itself does not have meaning. It's unfortunate, but the _compound noun_ for the set of content and structure we're talking about creating is **control-repo**.

If we're gonna rename it ...let's just say we can do WAY better than "control". If we're gonna call it what everybody calls it today though, we should use the full compound noun.

What we should not do is truncate the existing compound noun. Doing that doesn't give us any benefit and it will likely cause confusion.

If we stick with this name the CLI should be:
```
pdk new control-repo
```

**Comment by cdenneen** [comment]
Is anyone still working on this?

**Comment by scotje** [comment]
> Is anyone still working on this?

Yes! The very first control-repo related bits have landed recently (https://github.com/puppetlabs/pdk/pull/826) and we have been discussing this some at our bi-weekly public planning meetings.

There will be a more formal RFC for things soon, in the meantime you can see and comment on our working doc here: https://docs.google.com/document/d/1oNRjTvkkNSzuESlTazReOdYNgcdXo1Y-vV-LrA72-Cs/edit?usp=sharing

**Comment by bittner** [comment]
Any more progress to report on this topic?

I had an internal education and hands-on practicing with Puppet's PDK at our office yesterday, and I would have loved to show some related capabilities of PDK. You probably can tell the rest of the story yourselves... :roll_eyes: 

The feature descriptions in the linked Google doc are intriguing. I'd really love to see them hit the stage soon. :pray:
